src/DataStructures.jl

```diff
@@ -14,11 +14,11 @@ module DataStructures
                  ReverseOrdering, Reverse, Lt,
                  isless,
                  union, intersect, symdiff, setdiff, issubset,
-                 find, searchsortedfirst, searchsortedlast, endof, in
+                 find, searchsortedfirst, searchsortedlast, endof, in, produce
 
 
     export Deque, Stack, Queue
-    export deque, enqueue!, dequeue!, update!,iter
+    export deque, enqueue!, dequeue!, update!, forward_iter, backward_iter
```

Comment:
I modified things such that there are is a `DequeIterator` and a `ReverseDequeIterator` but the default iteration order is the `DequeIterator` for the deque object (i.e. if you just do
```
for i in dq
  println(i)
end
```

it iterates in the order you expect it to). I still think it makes sense to define a `forward_iter` method which can optionally be used to get objects in a FIFO ordering and a `reverse_iter` method which can optionally be used to get objects in a LIFO ordering. `Deque`, `Stack`, and `Queue` have their default iteration order defined as FIFO, LIFO, and FIFO respectively and in most cases people should just use their default iteration order (perhaps to make this clearer we should have the `iter` function which returns the default iteration order for the particular type - generally I think most people would use the conventional iterator syntax `for i in q` as opposed to `for i in forward_iter(q)`), but if you particularly want things in LIFO order then you would use `reverse_iter` (and vice versa for FIFO order).

Let me know if you think that's simply overcomplicating things. Also, regardless of the direction we take, we should document this behavior clearly.